### PR TITLE
Update code related to compiler attributes

### PR DESCRIPTION
**Pull request by memreflect**
tl;dr

* `unused` macro renamed to `UNUSED` for consistency with macros like `STMT` and `NOP`
* `noreturn(F)` macro and `noreturn` typedef consolidated into a single macro `Noreturn`
* New `FALLTHROUGH` macro to indicate explicit fallthrough to GCC-compatible compilers that will otherwise warn about it when a flag like `-Wimplicit-fallthrough` is enabled
* `UNUSED`, `Noreturn`, and `FALLTHROUGH` are now conditionally defined.

- - -

### `UNUSED` name change

I renamed `unused` to `UNUSED` since other macros like `STMT` and `NOP` use all-caps names.
I also used the attribute name `__unused__` in place of `unused` so that any potential `unused` macro already defined wouldn't expand `__attribute__((unused))` to something like `__attribute__((__attribute__((__unused__))))`.
I realize the name change is a slightly intrusive and pedantic change, but considering 10 files were already affected by other changes and i was already modifying nearby code in stdenv.h, another 4 files didn't seem that bad.

### `Noreturn` fixes

Since `USE_VOLATILE` was removed back in 0.9 beta1's switch to Autoconf, `#define volatile` is always used in `stdenv.h`, resulting in `noreturn` being equivalent to unqualified `void`.
After deleting the `#define volatile` line and its surrounding code, i got a lot of warnings from both GCC 11 and GCC 14 about `volatile void` thanks to `-Wignored-qualifiers` (enabled by `-W`).
Unless _es_ intends to support ancient versions of GCC prior to 3.0, there's no need for `volatile void` when the `noreturn` compiler attribute is available.

The separate `noreturn(Func)` macro and `noreturn` type definition were suspicious to me, so i combined them into a single `Noreturn` macro that expands to `__attribute__((__noreturn__)) void`.
This particular spelling also reflects its usage as a type name, like `Boolean`.

### `FALLTHROUGH`

I added a `FALLTHROUGH` macro expanding to `__attribute__((__fallthrough__))` to be used in place of fallthrough comments.
The macro expands to `(void)0` if not using a GCC-compatible compiler, allowing `FALLTHROUGH;` to be used portably.
With Clang, `-Wimplicit-fallthrough` is a simple on/off flag and is not enabled by `-W`, but explicit inclusion of `-Wimplicit-fallthrough` in Clang is like `-Wimplicit-fallthrough=5` in GCC, requiring the use of attributes to silence the compiler warnings (fallthrough comments are ineffective).
This is what actually led me to create a `FALLTHROUGH` macro.

### Miscellaneous changes/improvements

Conditional definitions are used for all three macros.
Someone using a C23 compiler might want `-DNoreturn='[[__noreturn__]] void'` for example, or some system header (e.g. `<sys/cdefs.h>` or similar) might have already defined an `UNUSED` macro that does the right thing for the system compiler instead of expanding to nothing if it doesn't claim to be a GCC-compatible compiler.

Speaking of no-return functions, C11 also introduced a `<stdnoreturn.h>` header that exists solely to `#define noreturn _Noreturn`.
If some system header innocuously included `<stdnoreturn.h>`, `PRIM(noreturn)` and `__attribute__((noreturn))` would result in a `$&_Noreturn` primitive in all compilers and `__attribute__((_Noreturn)) void` being used when compiling with a GCC-compatible compiler.
This is why i changed the name to `Noreturn` and used the `__noreturn__` attribute instead.
To avoid the `$&_Noreturn` issue, i also conditionally `#undef noreturn` just before `PRIM(noreturn)` to keep the primitive named `$&noreturn`.

`-Wimplicit-fallthrough=3` (enabled by `-W` in GCC) made GCC 11.4 think code after invocation of a function marked `Noreturn` would still be executed.
`signal.c` was affected with two uses of `NOTREACHED` occurring after invocations of `throw()` that would cause the compiler to erroneously warn about implicit fallthrough between cases in a `switch`.
While a minor annoyance that appears to have been fixed in a later release of GCC (GCC 14.2 for example), i still opted to remove the useless uses of `NOTREACHED` that caused the spurious warnings since GCC 11 is only 3 years old.

### CODE REVIEW

**Comment by jpco** [comment]
Very nice.  This is a clear improvement on how I did a lot of these macros.

Embarrassingly, I think the `volatile` change here explains why I wasn't having any success getting rid of the `-Wno-clobbered` flag...
